src/img/img-xpm.c

```diff
@@ -0,0 +1,1968 @@
+		return NULL;
+	}
+	if (w <= 0) {
+		return NULL;
+	}
+	if (h <= 0) {
+		return NULL;
+	}
+	if (cpp <= 0 || cpp >= 32) {
+		return NULL;
+	}
+	if (n_col <= 0 || n_col >= INT_MAX / (cpp + 1)
+			|| n_col >= INT_MAX / (int)sizeof(struct xpm_color)) {
+		return NULL;
+	}
+
+	name_buf = (char *)calloc(n_col, cpp + 1);
+	if (!name_buf) {
+		return NULL;
+	}
+	colors = (struct xpm_color *)calloc(n_col, sizeof(struct xpm_color));
+	if (!colors) {
+		free(name_buf);
+		return NULL;
+	}
+
+	for (cnt = 0; cnt < n_col; cnt++) {
+		char *color_name;
+
+		buffer = (*get_buf)(op_cmap, handle);
+		if (!buffer) {
+			free(colors);
+			free(name_buf);
+			return NULL;
+		}
+
+		color = &colors[cnt];
+		color->color_string = &name_buf[cnt * (cpp + 1)];
+		strncpy(color->color_string, buffer, cpp);
+		color->color_string[cpp] = 0;
+		buffer += strlen(color->color_string);
+		color->transparent = 0;
+
+		color_name = xpm_extract_color(buffer);
+
+		if (!color_name || (strcasecmp(color_name, "None") == 0) ||
+				(parse_color(color_name, color) == 0)) {
+			color->transparent = 1;
+			color->red = 0;
+			color->green = 0;
+			color->blue = 0;
+			is_trans = 1;
+		}
+
+		free(color_name);
+
+		if (cnt == 0) {
+			fallbackcolor = color;
+		}
+	}
+
+	data = (uint32_t *)calloc(w * h, sizeof(uint32_t));
+	if (!data) {
+		free(colors);
+		free(name_buf);
+		return NULL;
+	}
+
+	wbytes = w * cpp;
+
+	for (ycnt = 0; ycnt < h; ycnt++) {
+		buffer = (*get_buf)(op_body, handle);
+		if (!buffer || ((int)strlen(buffer) < wbytes)) {
+			continue;
+		}
+
+		for (n = 0, xcnt = 0; n < wbytes; n += cpp, xcnt++) {
+			uint32_t a, r, g, b;
+
+			strncpy(pixel_str, &buffer[n], cpp);
+			pixel_str[cpp] = 0;
+
+			color = lookup_color(colors, n_col, pixel_str);
+
+			/* Bad XPM...punt */
+			if (!color) {
+				color = fallbackcolor;
+			}
+
+			a = 0xFF;
+			if (is_trans && color->transparent) {
+				a = 0;
+			}
+			r = color->red >> 8;
+			g = color->green >> 8;
+			b = color->blue >> 8;
+			data[ycnt * w + xcnt] = (a << 24) | (r << 16) | (g << 8) | b;
+		}
+	}
+
+	free(colors);
+	free(name_buf);
+
+	*width = w;
+	*height = h;
+	return data;
+}
+
+void
+img_xpm_load(const char *filename, struct lab_data_buffer **buffer)
+{
+	if (*buffer) {
+		wlr_buffer_drop(&(*buffer)->base);
+		*buffer = NULL;
+	}
+	if (string_null_or_empty(filename)) {
+		return;
+	}
+
+	struct file_handle handle;
+	int w, h;
+	uint32_t *data;
+	cairo_surface_t *surface;
+	unsigned char *surface_data;
+
+	memset(&h, 0, sizeof(h));
+	handle.infile = fopen(filename, "r");
+	if (!handle.infile) {
+		fprintf(stderr, "Failed to load XPM file: %s\n", filename);
+		return;
+	}
+	handle.buffer_size = 4096;
+	handle.buffer = calloc(4096, 1);
+	data = pixbuf_create_from_xpm(file_buffer, &handle, &w, &h);
+	free(handle.buffer);
+	fclose(handle.infile);
+	if (!data) {
+		fprintf(stderr, "Failed to load XPM file: %s\n", filename);
+		return;
+	}
+
+	surface = cairo_image_surface_create(CAIRO_FORMAT_ARGB32, w, h);
```

Comment:
Is there an advantage to using a cairo surface? I just used buffer_create_wrap() with the raw ARGB data and it seemed to work all right...

Reply:
I think `buffer_create_wrap()` is better than doing a cairo surface.
Aside... we really ought to rename `buffer_create_wrap()`!